plugin/storage/es/samplingstore/storage.go

```diff
@@ -0,0 +1,189 @@
+		Hostname:      hostname,
+		Probabilities: probabilities,
+		QPS:           qps,
+	}
+	s.writeProbabilitiesAndQPS(writeIndexName, ts, val)
+	return nil
+}
+
+func (s *SamplingStore) writeProbabilitiesAndQPS(indexName string, ts time.Time, pandqps dbmodel.ProbabilitiesAndQPS) {
+	s.client().Index().Index(indexName).Type(probabilitiesType).
+		BodyJson(&dbmodel.TimeProbabilitiesAndQPS{
+			Timestamp:           ts,
+			ProbabilitiesAndQPS: pandqps,
+		}).Add()
+}
+
+func indexWithDate(indexNamePrefix, indexDateLayout string, date time.Time) string {
+	return indexNamePrefix + date.UTC().Format(indexDateLayout)
+}
+
+func (s *SamplingStore) writeThroughput(indexName string, ts time.Time, throughputs []*model.Throughput) {
+	s.client().Index().Index(indexName).Type(throughputType).
+		BodyJson(&dbmodel.TimeThroughput{
+			Timestamp:  ts,
+			Throughput: dbmodel.FromThroughputs(throughputs),
+		}).Add()
+}
+
+func (s *SamplingStore) GetThroughput(start, end time.Time) ([]*model.Throughput, error) {
+	ctx := context.Background()
+	indices := s.getReadIndices(start, end)
+	searchResult, err := s.client().Search(indices...).
+		Size(s.maxDocCount).
+		Query(buildTSQuery(start, end)).
+		IgnoreUnavailable(true).
+		Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for throughputs: %w", err)
+	}
+	var retSamples []dbmodel.Throughput
+	hits := searchResult.Hits.Hits
+	for _, hit := range hits {
+		source := hit.Source
+		var tToD dbmodel.TimeThroughput
+		if err := json.Unmarshal(*source, &tToD); err != nil {
+			return nil, errors.New("unmarshalling ElasticSearch documents failed")
+		}
+		retSamples = append(retSamples, tToD.Throughput...)
+	}
+	return dbmodel.ToThroughputs(retSamples), nil
+}
+
+func (s *SamplingStore) GetLatestProbabilities() (model.ServiceOperationProbabilities, error) {
+	ctx := context.Background()
+	indices := s.getLatestIndices()
+	searchResult, err := s.client().Search(indices...).
+		Size(s.maxDocCount).
+		IgnoreUnavailable(true).
+		Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for Latest Probabilities: %w", err)
+	}
+	lengthOfSearchResult := len(searchResult.Hits.Hits)
+	if lengthOfSearchResult == 0 {
+		return nil, nil
+	}
+
+	var latestProbabilities dbmodel.TimeProbabilitiesAndQPS
+	latestTime := time.Time{}
+	for _, hit := range searchResult.Hits.Hits {
+		var unMarshalProbabilities dbmodel.TimeProbabilitiesAndQPS
+		err = json.Unmarshal(*hit.Source, &unMarshalProbabilities)
+		if err != nil {
+			return nil, err
+		}
+		if unMarshalProbabilities.Timestamp.After(latestTime) {
+			latestTime = unMarshalProbabilities.Timestamp
+			latestProbabilities = unMarshalProbabilities
+		}
+	}
+	return latestProbabilities.ProbabilitiesAndQPS.Probabilities, nil
+}
+
+func buildTSQuery(start, end time.Time) elastic.Query {
+	return elastic.NewRangeQuery("timestamp").Gte(start).Lte(end)
+}
+
+func (s *SamplingStore) getLatestIndices() []string {
+	currTime := time.Now().UTC()
```

Comment:
this doesn't look good. If a user rotates indices every hour, and for some reason the adaptive sampling hasn't run for the last hour, this will never find the records that may actually be in the database. Maybe it needs some kind of lookback configuration parameter.

Reply:
added an algorithm in recent commit to get latest existing index.